### PR TITLE
POLIO-1037 reasons delay tracking

### PR DESCRIPTION
**Pull request by quang-le**
Add a feature to keep track of the polio campaign round date changes, the users who perform those changes and the reason for the changes

Related JIRA tickets :  POLIO-1037

## Self proofreading checklist

- [X] Did I use eslint and black formatters
- [X] Is my code clear enough and well documented
- [X] Are my typescript files well typed
- [X] New translations have been added or updated if new strings have been introduced in the frontend
- [X] My migrations file are included
- [ ] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Changes

1. Back-end
- Add `RoundDateHistory` Model, Serializer, Viewset and admin entry. The endpoint is `/api/polio/datelogs/`. It only allows get requests and is used by the front-end with the `round__id` param to display the history of a given round
- The `RoundDateHistorySerializer` has a `validate` method that performs the same checks as the front-end validation, except comparing the dates with the dates of the previous and next round.
- Update `RoundSerializer` to handle the `datelogs` field, which regroups the `RoundDateHistoryEntry`s for a round
- Hard-code 2 possible choices for 'reason` for date change: initial data and encoding error. Other will be added when we get feedback from the client
- Add migration to automatically add an "initial date" datelog entry for all campaigns. These entries are easily recognizable as they are not associated to any user

2. Front-end
- Replaced the `DateInput`s for round start date and end date with `RoundDates` component. If there are no start or end date, it will display the `DateInput`, and the form will automatically add an entry to the `datelogs` with initial_data` as `reason`. If there's at least one date, the `DateInput`s will be replaced by text, with an `Edit` button and an `Info`  icon button
- The IconButton opens a modal (new `AlertModal` component, with just a `Close` button) which will display a table with all the `datelogs` for the round
- The `Edit` button will open a modal where users can change  the dates and add a reason for date change. The reason for date change is mandatory and users need to modify at least one date to be able to confirm. The date fields are `DateInput`s with the same validation as the current validation in place for round dates 
- Did some cleanup and TS conversion

## UPDATE 09/06/2023
- Add datelogs to campaign csv export

## How to test

1. Front-end
- Go to polio>campaigns
- Create a new campaign, add an OBR name and an initial region
- Go to Rounds tab
- Try filling out dates and play with the validation: end date before start date an vice versa, dates for round 2 that are earlier than those of round 1 etc
- Settle for correct date and save
- Re-open the campaign
- Try editing the dates
- Play with the dates as earlier to test the validation
- Save again
- Repeat until you have made and saved at least 6 date changes
- open the campaign again
- click on the icon button to see the history
- The table should display all you changes
- Play around with the table: only created_at and modified_by should be sortable. Test the pagination and the results displayed

2. Backend
- Copy the id of your test campaign
- Go to `/api/polio/campaigns/<id_of_test_campaign>`
- Copy the json response
- In the Django API page, paste the json in the content section. 
- In the json, find round with datelogs
- Add a datelog with faulty dates (same sort as front-end) and click `PUT`
- You should get validation error messages

## Print screen / video

https://github.com/BLSQ/iaso/assets/38907762/a8aa66ba-6e5d-489e-a670-a847db2690b5

## Notes

Created ticket https://bluesquare.atlassian.net/browse/IA-2234 to move `AlertModal` to bluesquare-components

@olethanh & @mathvdh  please have a good look at the manual migration

### CODE REVIEW

File: plugins/polio/api.py [hunk 1]
```diff
@@ -1735,6 +1737,22 @@ def get_queryset(self):
         return Config.objects.filter(users=self.request.user)
 
 
+@swagger_auto_schema(tags=["datelogs"])
+class RoundDateHistoryEntryViewset(ModelViewSet):
+    http_method_names = ["get"]
+    permission_classes = [permissions.IsAuthenticatedOrReadOnly]
+    serializer_class = RoundDateHistoryEntrySerializer
+    queryset = RoundDateHistoryEntry.objects.all()
```
Comment (olethanh) [inline plugins/polio/api.py]: to respect proper permission this  should have a get_queryset method that limit the result to the campaign the user has access to

File: plugins/polio/admin.py [hunk 2]
```diff
@@ -128,11 +129,16 @@ class SurgeAdmin(admin.ModelAdmin):
     formfield_overrides = {models.JSONField: {"widget": IasoJSONEditorWidget}}
 
 
+class RoundDateHistoryEntryAdmin(admin.ModelAdmin):
```
Comment (mathvdh) [inline plugins/polio/admin.py]: This is not needed if you do nothing in it ...

File: plugins/polio/admin.py [hunk 3]
```diff
@@ -128,11 +129,16 @@ class SurgeAdmin(admin.ModelAdmin):
...
+    pass
+
+
 admin.site.register(Campaign, CampaignAdmin)
 admin.site.register(CampaignGroup, CampaignGroupAdmin)
 admin.site.register(Config, ConfigAdmin)
 admin.site.register(Surge, SurgeAdmin)
 admin.site.register(Round)
+admin.site.register(RoundDateHistoryEntry, RoundDateHistoryEntryAdmin)
```
Comment (mathvdh) [inline plugins/polio/admin.py]: You can just write `admin.site.register(RoundDateHistoryEntry)`

File: plugins/polio/api.py [hunk 4]
```diff
@@ -1735,6 +1737,26 @@ def get_queryset(self):
...
```
Comment (mathvdh) [inline plugins/polio/api.py]: A mon avis tu peux virer ça et tout mettre dans `get_queryset`

File: plugins/polio/models.py [hunk 5]
```diff
@@ -74,6 +74,8 @@
     ("DFC", _("DFC")),
 ]
 
+DELAY_REASONS = [("INITIAL_DATA", _("initial_data")), ("ENCODING_ERROR", _("encoding_error"))]
```
Comment (mathvdh) [inline plugins/polio/models.py]: Olivier propose qu'on utilise le django builtin TextChoice : https://docs.djangoproject.com/en/3.2/ref/models/fields/#enumeration-types

**Review by mathvdh — CHANGES_REQUESTED** [review]
I only did the python part and I don't understand about this part of the project, but still some small modification ...

File: plugins/polio/api.py (hunk 4 above)
Comment (quang-le) [reply plugins/polio/api.py]: J'avoue j'ai regardé comment c'était fait ailleurs et j'ai fait pareil

Comment (quang-le) [reply plugins/polio/api.py]: j'ai viré `queryset` mais laissé le reste

**Review by mathvdh — APPROVED** [review]
👍🏼

**Comment by hakifran** [comment]
@quang-le while testing I'm getting this error when saving changes on dates: {
 "name": "ApiError",
 "status": 500,
 "message": "Error on PUT /api/polio/campaigns/7df09084-63cb-434d-bf6d-fada9f3a6667/ "
}
Screenshot from 2023-06-13 14-17-12

**Review by hakifran — CHANGES_REQUESTED** [review]
There is a bug when the campaign is edited with round 0

**Review by hakifran — APPROVED** [review]

**Pull request merged**